### PR TITLE
Bugs fixes And Improvements of MBE

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR fixed / improve followings things:

* Add `token` `$repeat` separator support: Previously $repeat only support single punct separator. 
* Fixed a bug which expand infinite pattern, see `test_match_group_in_group` 
* Correctly handle +,*,? case of $repeat patterns
* Increase the limit of $repeat patterns (128 => 65536), personally i think we could remove this limit as we seem to fix all major loop bugs
* **Re-enable tt matcher**
* Better meta item parsing.
* Add related tests and add some real world test cases.
* Add more debug information.

### CODE REVIEW

File: crates/ra_mbe/src/subtree_source.rs [hunk 1]
```diff
@@ -437,12 +437,16 @@ fn convert_delim(d: tt::Delimiter, closing: bool) -> TtToken {
...
-        is_joint_to_next: false,
-        text: l.text.clone(),
-        n_tokens: 1,
-    }
+    let kind = classify_literal(&l.text)
+        .map(|tkn| tkn.kind)
+        .or_else(|| match l.text.as_ref() {
+            "true" => Some(SyntaxKind::TRUE_KW),
+            "false" => Some(SyntaxKind::FALSE_KW),
+            _ => None,
```
Comment (matklad) [inline crates/ra_mbe/src/subtree_source.rs]: Let's panic here, instead of `unwrap`

File: crates/ra_parser/src/grammar.rs [hunk 2]
```diff
@@ -119,7 +119,22 @@ pub(crate) fn meta_item(p: &mut Parser) {
...
+                    p.bump();
+                    match p.current() {
+                        c if c.is_literal() => p.bump(),
+                        TRUE_KW | FALSE_KW => p.bump(),
+                        _ => {}
+                    }
+                    break;
+                }
+                _ => break,
+            }
```
Comment (matklad) [inline crates/ra_parser/src/grammar.rs]: Let's add a test for it?

Also, this was recently changed in https://github.com/rust-lang/rust/pull/57367/, I am not sure if the refercne is up-to-date

Comment (edwin0cheng) [reply crates/ra_parser/src/grammar.rs]: I don't know where to add the test, as this function only called from mbe right now.
I can add that test in mbe if you want.

Comment (matklad) [reply crates/ra_parser/src/grammar.rs]: Ah, right, the parser doesn't need this itself. I guess I am fine with existing tests in mbe then!

**Comment by edwin0cheng** [comment]
Finally i feel confident to refactor the lexer and mbe with enough test case :)

**Comment by matklad** [comment]
@edwin0cheng yeah, I also feel that that's just the right approach: build the basic test-suite using simplistic solution, then refactor to the proper infra, and then the remaining 90% of fixing all of the bugs :D

**Pull request merged**